include/zephyr/drivers/counter.h

```diff
@@ -210,35 +227,106 @@ struct counter_config_info {
+ */
+struct counter_top_cfg_64 {
+	/**
+	 * Top value.
+	 */
+	uint64_t ticks;
+	/**
+	 * Callback function (can be NULL).
+	 */
+	counter_top_callback_t callback;
+	/**
+	 * User data passed to callback function (not valid if callback is NULL).
+	 */
+	void *user_data;
+	/**
+	 * Flags (see @ref COUNTER_TOP_FLAGS).
+	 */
+	uint32_t flags;
+};
+
 typedef int (*counter_api_start)(const struct device *dev);
 typedef int (*counter_api_stop)(const struct device *dev);
-typedef int (*counter_api_get_value)(const struct device *dev,
-				     uint32_t *ticks);
-typedef int (*counter_api_get_value_64)(const struct device *dev,
-			uint64_t *ticks);
+typedef int (*counter_api_get_value)(const struct device *dev, uint32_t *ticks);
 typedef int (*counter_api_reset)(const struct device *dev);
-typedef int (*counter_api_set_alarm)(const struct device *dev,
-				     uint8_t chan_id,
+typedef int (*counter_api_set_value)(const struct device *dev, uint32_t ticks);
+typedef int (*counter_api_set_alarm)(const struct device *dev, uint8_t chan_id,
 				     const struct counter_alarm_cfg *alarm_cfg);
-typedef int (*counter_api_cancel_alarm)(const struct device *dev,
-					uint8_t chan_id);
+typedef int (*counter_api_cancel_alarm)(const struct device *dev, uint8_t chan_id);
 typedef int (*counter_api_set_top_value)(const struct device *dev,
 					 const struct counter_top_cfg *cfg);
 typedef uint32_t (*counter_api_get_pending_int)(const struct device *dev);
 typedef uint32_t (*counter_api_get_top_value)(const struct device *dev);
-typedef uint32_t (*counter_api_get_guard_period)(const struct device *dev,
-						 uint32_t flags);
-typedef int (*counter_api_set_guard_period)(const struct device *dev,
-						uint32_t ticks,
-						uint32_t flags);
+typedef uint32_t (*counter_api_get_guard_period)(const struct device *dev, uint32_t flags);
+typedef int (*counter_api_set_guard_period)(const struct device *dev, uint32_t ticks,
+					    uint32_t flags);
 typedef uint32_t (*counter_api_get_freq)(const struct device *dev);
+typedef uint64_t (*counter_api_get_freq_64)(const struct device *dev);
+
+typedef int (*counter_api_get_value_64)(const struct device *dev, uint64_t *ticks);
+typedef int (*counter_api_set_value_64)(const struct device *dev, uint64_t ticks);
+typedef int (*counter_api_set_alarm_64)(const struct device *dev, uint8_t chan_id,
+					const struct counter_alarm_cfg_64 *alarm_cfg);
+typedef uint64_t (*counter_api_get_guard_period_64)(const struct device *dev, uint32_t flags);
+typedef int (*counter_api_set_guard_period_64)(const struct device *dev, uint64_t ticks,
+					       uint32_t flags);
+typedef uint64_t (*counter_api_get_top_value_64)(const struct device *dev);
+typedef int (*counter_api_set_top_value_64)(const struct device *dev,
+					    const struct counter_top_cfg_64 *cfg);
 
 __subsystem struct counter_driver_api {
 	counter_api_start start;
 	counter_api_stop stop;
 	counter_api_get_value get_value;
-	counter_api_get_value_64 get_value_64;
```

Comment:
This field now being gated by a Kconfig means we still have a breaking change for downstream drivers (as illustrated by the update required for in-tree drivers)

Reply:
then it will just be added to the migration guide

Reply:
> yeah... we already had that discussion in the arch meeting. This API is not technically stable, and changes can still happen.
> 
> I added a note to the migration guide for 4.4

I believe the conclusion of the discussion was that we actually wanted to treat it as stable, and as a matter of fact it now is https://github.com/zephyrproject-rtos/zephyr/pull/98919